tutorials/nservicebus-101/lesson-2/tutorial.md

```diff
@@ -0,0 +1,191 @@
+
+snippet:EmptyHandler
+
+The implementation of the `IHandleMessages<T>` interface is the `Handle` method, which NServiceBus will invoke when a matching message (in this case `DoSomething`) arrives. The `Handle` method receives the message and an `IMessageHandlerContext` that contains tools for working with messages.
+
+The `Handle` method must return a [`System.Threading.Tasks.Task`](https://msdn.microsoft.com/en-us/library/system.threading.tasks.task.aspx). Returning `null` will result in an exception. In the case above, no async calls need to be made, so we return `Task.CompletedTask` to fulfill the return requirement. 
+
+NOTE: Before .NET 4.6, `Task.CompletedTask` is not available. In that case, you can use `Task.FromResult(false)` instead.
+
+Instead, we can add the `async` keyword to the method, and then the compiler takes care of the tasks more or less transparently. No explicit return is needed because the compiler generates it for you as part of the async state machine.
+
+snippet:EmptyHandlerAsync
+
+You can even implement `IHandleMessages<T>` for multiple different message types within the same class, if it makes sense to logically group the handlers together. Just don't expect any state in class variables to be persisted. NServiceBus will create a new instance of the class for every message that it handles.
+
+snippet:MultiHandler
+
+When NServiceBus starts up, it will find all of these message handler classes and automatically wire them up, so that they will be invoked when messages arrive. There's no special setup or configuration needed.
+
+
+## Exercise
+
+Now let's take the "Hello world" solution we started in the last lesson and modify it to send a message. You can also grab the [completed solution from the last lesson](https://github.com/Particular/docs.particular.net/tree/academy-nsb101/tutorials/nservicebus-101/lesson-1/solution/) to use as a starting point.
+
+When we're done, the ClientUI endpoint will be sending a PlaceOrder message to itself, and then processing that message, as depicted in the following diagram:
+
+![Exercise 2 Diagram](diagram.png)
+
+
+### Create a messages assembly
+
+Messages should be self-contained within a separate assembly so that they can be shared between endpoints, so let's create that assembly now.
+
+1. In the solution, create a new project and select the **Class Library** project type.
+2. Set the name of the project to **Messages**.
+3. Remove the automatically created **Class1.cs** file from the project. We would prefer to have our messages a little more organized than files in the root of the project. (More on this a little later.)
+4. Add the NServiceBus NuGet package to the Messages project.
+5. In the **ClientUI** project, add a reference to the **Messages** project.
+
+
+### Create a message
+
+In order to better organize messages, we'll create our first command in a folder called **Commands**.
```

Comment:
this is actually an antipattern. messages should be grouped by requirement not by Commands/events etc